### ISSUE TITLE
Bug: [require-array-sort-compare] false positive on a type parameter constrained to a string array

### ISSUE THREAD

**Issue by zaewc** [typescript-eslint/typescript-eslint#12511]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.3&fileType=.tsx&code=GYVwdgxgLglg9mABMAPAFUQUwB5U2AEwGdEioAnGMAcwG0BdAPgAoBDALkTQEpEBvAFCJErAHRE45KM24BuAQF8BQA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwo2V4A8pTIB7BogxhgveCIBWyYjkmQAdMxwkxAMzIBzVXMRjIAX24RI8AK60ik4JFwEipCtTqMW8ZAEdrZD5UCPCweFTG8NTEIgC2bMgqqLJopiBpQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
function f<T extends string[]>(a: T) {
  a.sort();
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  parserOptions: { project: "./tsconfig.json" },
  rules: {
    "@typescript-eslint/require-array-sort-compare": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error

`ignoreStringArrays` defaults to `true`, and the elements are all strings, same as `function f(a: string[]) { a.sort(); }`, which isn't reported.

### Actual Result

Require 'compare' argument. 2:3 - 2:11

**Comment by StyleShit** [comment]
Yup, definitely looks like a bug

Weirdly enough, this case is not reported:

```TS
function f<T extends string>(a: T[]) {
  a.sort();
}
```
